providers/implementations/macs/blake2_mac_impl.c

```diff
@@ -87,15 +87,25 @@ static size_t blake2_mac_size(void *vmacctx)
     return macctx->params.digest_length;
 }
 
-static int blake2_mac_init(void *vmacctx)
+static int blake2_mac_init(void *vmacctx, const unsigned char *key,
+                           size_t keylen, const OSSL_PARAM params[])
 {
     struct blake2_mac_data_st *macctx = vmacctx;
 
-    if (!ossl_prov_is_running())
+    if (!ossl_prov_is_running() || !blake2_mac_set_ctx_params(macctx, params))
         return 0;
-
-    /* Check key has been set */
-    if (macctx->params.key_length == 0) {
+    if (key != NULL) {
+        if (keylen > BLAKE2_KEYBYTES || keylen == 0) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH);
+            return 0;
+        }
+        memcpy(macctx->key, key, keylen);
+        /* Pad with zeroes at the end if required */
+        if (keylen < BLAKE2_KEYBYTES)
```

Comment:
Not sure if there is a way of sharing this block with the settable case..

Reply:
There is.